### ISSUE TITLE
Add feature: auto-height

### ISSUE THREAD

**Issue by GitHubish** [MurhafSousli/ngx-scrollbar#214]
Bug Report or Feature Request (mark with an x)
- [x] bug report

OS and Version?  Windows 10 64bits - Chrome 78

Versions
Angular CLI: 8.1.1
Node: 10.16
Angular: 8.1.1
Ngx-Scrollbar: 6.0.0

Repro steps
HTML
`<ng-scrollbar #scrollable track="all" style="max-height: calc(100% - 25px);">`

Desired functionality
It seems that when I use a maxHeight defined in pixels (for example: 300px) or with a calculation in css in relation to the viewport or relative to its parent (ex: calc(100% - 25px)) it doesn't work.

reproduction: [stackblitz](https://ngx-scrollbar-j22yim.stackblitz.io/#/)

The example shows a version with a native scrollbar, where the max-height works. While with NGX-SCROLLBAR the block is 0px high. It does not support max-height. How to do this?

Thanks

**Comment by Derexx** [comment]
Same error to me after upgrade to latest version.

**Comment by MurhafSousli** [comment]
There is no error, the height of `<ng-scrollbar>` must be set. if you want to use percent value for the height then the parent must have a fixed height

This would be a feature request

**Comment by GitHubish** [comment]
Yes, totally it is a request for functionality. However, between version 4 and version 6 I do have a change of behavior, result I have a scrollbar with a height at 0 whereas before it took all the height.

Do you think this feature can be added to manage the height in percentage and in relation to the viewport? (ex: calc(100vh - 200px)
thanks

**Comment by chiraggmodi** [comment]
same issue, I want to assign max-height, as my content height is not fixed, it may be less then height I assign in css, so if the content is less, then there will more blank space in content.

**Comment by Nekronik** [comment]
Hello @MurhafSousli I'm also facing this.

As stated before, in earlier versions, the height of a scrollbar was the same as the content and to decide wether or not display a scroll a max-height was used.
The current behavior forces the user to always set a height even if the content is not always the same or it changes over time. This behavior applies even if the direction of the scrollbar is set as `horizontal`.

In order to update to version 6 I have to calculate a height for something that always has different height and might change, and not only that, I have to do this for tens of components.

:question:  I (and I'm sure many other) would like to know if this change happened because it was the "only way to proceed" or if the old behavior could be restored with some effort or a pull request.

As always thanks :smiley:

**Comment by ruizmarc** [comment]
I'm joining this feature request. It would be very helpful to get the scrollbar work with max-height... otherwise I need to manually calculate the height of a lot of components and recalculate height on resize...

**Comment by MurhafSousli** [comment]
Hi @ruizmarc, I am currently working on a new feature for v7.0.0. so I am doing some refactoring.
Well basically all that we need is to set `<ng-scrollbar>` height to the content wrapper height

```html
 <ng-scrollbar #scrollable (updated)="autoHeight()">
</ng-scrollbar>
```
```ts
import { Component, ViewChild, ElementRef } from "@angular/core";
import { NgScrollbar } from "ngx-scrollbar";

@Component({
  selector: "app-home",
  templateUrl: "./home.component.html",
  styleUrls: ["./home.component.scss"]
})
export class HomeComponent {

  @ViewChild("scrollable", { static: true, read: ElementRef }) scrollbar: ElementRef;

  autoHeight() {
    const contentWrapper = document.querySelector(".ng-scroll-content");
    if (this.scrollbar) {
      this.scrollbar.nativeElement.style.height =
        contentWrapper.clientHeight + "px";
    }
  }
}
```

Fixed stackblitz https://stackblitz.com/edit/ngx-scrollbar-gap4cs?file=src/app/home/home.component.html

I will add this in the next version

**Comment by GitHubish** [comment]
Great if a new version is on the way with this support! 
On the fix you published, the scrollbar is not functional for me.

**Comment by MurhafSousli** [comment]
Please try v7 and let me know if you still have the issue

**Comment by Nekronik** [comment]
Hello @MurhafSousli 

I have just tested the autoHeight and it is not working as it is.

After some digging arround I got to the conclusion that the `autoHeightStyles` need to be set to `ng-scrollbar` instead of `ng-scrollbar-wrapper`.

Setting the styles to the root allow the children (which use 100% values) to inherit it.
Doing so makes the scrollbar not appear unless a size restriction (`max-height` or `max-width`) is set, and this alignes with native.

While we are at it. It would be nice that `ng-scrollbar-wrapper` and `ng-scroll-viewport-wrapper` inherited the `border-radius` for prettier scrollbar containers. (I don't know if it can be set from outside)

Good work :+1:

**Comment by MurhafSousli** [comment]
> I have just tested the autoHeight and it is not working as it is.

@Nekronik can you explain what is not working exactly? maybe provide a stackblitz if possible!


The only problem I found with the current auto-height implementation is setting the proper width with `mat-dialog` component.

```ts
matDialog.open(mySrollableComponent);
```
but it works properly if you set the dialog width e.g. 

```ts
matDialog.open(mySrollableComponent, { width: '300px'});
```

---

I use `ng-scrollbar-wrapper` as the functional layer, I prefer not to use the root element because a user might bind its style, e.g. `<ng-scrollbar [style.height]="someValue">`

Anyway, I find no difference when I use the root element `ng-scrollbar` to set the `autoHeightStyles`, try the apply the following [gist](https://gist.github.com/MurhafSousli/9eae8d569572da8d94d0911cd3cb142f) and see if it fixes your issue

**Comment by MurhafSousli** [comment]
@Nekronik Now it is much better with a repro, here is my attempt to fix it in this [stackblitz](https://stackblitz.com/edit/ngx-scrollbar-compare-native-auto-height-qmm8ug?file=src/app/comparison/comparison/comparison.component.ts)

There is still a problem if `ng-scrollbar` got margin, therefore `min-width` and `max-width` should be set manually

```css
ng-scrollbar {
   margin: 1em;
    // if margin is used, it must use a calculated width to get the proper behavior.
    min-width: calc(100% - 2em); 
    max-width: calc(100% - 2em); 
}
```
Any idea how to solve this?

**Comment by MurhafSousli** [comment]
@Nekronik 
> While we are at it. It would be nice that ng-scrollbar-wrapper and ng-scroll-viewport-wrapper inherited the border-radius for prettier scrollbar containers. (I don't know if it can be set from outside)

As I see, the border-radius looks fine to me in your stackblitz!

It is possible to set `ng-scrollbar-wrapper` and `ng-scroll-viewport-wrapper` border-radius from outside using a CSS variable, but how does it make a difference? isn't setting the border-radius of `ng-scrollbar` enough?

**Comment by MurhafSousli** [comment]
Ok, it is a bit tricky, When we set the host component width and height, it affects the viewport element dimension as well, which causes the resize observer to update the host component again, and we get an infinite update loop.

Do you have any idea?

**Comment by MurhafSousli** [comment]
Fixed in v7.0.0, [auto-height stackblitz](https://stackblitz.com/edit/ngx-scrollbar-compare-native-auto-height-a3akqi?file=src%2Fapp%2Fcomparison%2Fcomparison%2Fcomparison.component.ts)

**Comment by lozah** [comment]
Hello, 

There is an other problem with the auto-height behaviour.
When an element is closed and then opened the scrollbar appears even if there is enough space to display all the content. The scrollbar disappears on hover over the element.

I reproduced the problem on your auto-height stackbitz to show you. 

[stackbiltz](https://stackblitz.com/edit/ngx-scrollbar-compare-native-auto-height-f7ax8n)

**Comment by TejashreeD** [comment]
@GitHubish The fix you given in the question itself.`style="max-height: calc(100% - 25px);"` is worked for me. Thanks a lot. 

`    "ngx-scrollbar": "^7.2.3",
    "@angular/cdk": "^9.2.4",
    "@angular/cli": "~9.0.6",
`

**Comment by dstj** [comment]
In case someone else stumbles here like I did:

> Fixed in v7.0.0, [auto-height stackblitz](https://stackblitz.com/edit/ngx-scrollbar-compare-native-auto-height-a3akqi?file=src%2Fapp%2Fcomparison%2Fcomparison%2Fcomparison.component.ts)

As per [this comment](https://github.com/MurhafSousli/ngx-scrollbar/issues/356#issuecomment-818237364), since 7.5.0, `[autoHeightDisabled]` is set to `true` by default, so you must now set `[autoHeightDisabled]="false"` for auto-height to work.

**Comment by keyone-zen** [comment]
for anyone else that has this issue

```
<ng-scrollbar [autoHeightDisabled]="false">
  <div scrollViewport mat-dialog-content>
  </div>
</ng-scrollbar>
```
